iris-mpc-cpu/src/network/value.rs

```diff
@@ -16,20 +17,231 @@ pub enum NetworkValue {
+                    u64::from_le_bytes(<[u8; 8]>::try_from(&serialized[1..9])?),
+                )))
+            }
+            0x06 => {
+                if serialized.len() < 5 {
+                    return Err(eyre!(
+                        "Invalid length for VecRing16: can't parse vector length"
+                    ));
+                }
+                let len = u32::from_le_bytes(<[u8; 4]>::try_from(&serialized[1..5])?) as usize;
+                if serialized.len() != 5 + 2 * len {
+                    return Err(eyre!("Invalid length for VecRing16"));
+                }
+                let mut res = Vec::with_capacity(len);
+                for i in 0..len {
+                    res.push(RingElement(u16::from_le_bytes(<[u8; 2]>::try_from(
+                        &serialized[5 + 2 * i..5 + 2 * (i + 1)],
+                    )?)));
+                }
+                Ok(NetworkValue::VecRing16(res))
+            }
+            0x07 => {
+                if serialized.len() < 5 {
+                    return Err(eyre!(
+                        "Invalid length for VecRing32: can't parse vector length"
+                    ));
+                }
+                let len = u32::from_le_bytes(<[u8; 4]>::try_from(&serialized[1..5])?) as usize;
+                if serialized.len() != 5 + 4 * len {
+                    return Err(eyre!("Invalid length for VecRing32"));
+                }
+                let mut res = Vec::with_capacity(len);
+                for i in 0..len {
+                    res.push(RingElement(u32::from_le_bytes(<[u8; 4]>::try_from(
+                        &serialized[5 + 4 * i..5 + 4 * (i + 1)],
+                    )?)));
+                }
+                Ok(NetworkValue::VecRing32(res))
+            }
+            0x08 => {
+                if serialized.len() < 5 {
+                    return Err(eyre!(
+                        "Invalid length for VecRing64: can't parse vector length"
+                    ));
+                }
+                let len = u32::from_le_bytes(<[u8; 4]>::try_from(&serialized[1..5])?) as usize;
+                if serialized.len() != 5 + 8 * len {
+                    return Err(eyre!("Invalid length for VecRing64"));
+                }
+                let mut res = Vec::with_capacity(len);
+                for i in 0..len {
+                    res.push(RingElement(u64::from_le_bytes(<[u8; 8]>::try_from(
+                        &serialized[5 + 8 * i..5 + 8 * (i + 1)],
+                    )?)));
+                }
+                Ok(NetworkValue::VecRing64(res))
+            }
+            _ => Err(eyre!("Invalid network value type")),
+        }
     }
 
     pub fn vec_to_network(values: &Vec<Self>) -> Vec<u8> {
-        bitcode::serialize(&values).unwrap()
+        let mut res = vec![];
+        res.extend_from_slice(&(values.len() as u32).to_le_bytes());
+        for value in values {
+            res.extend_from_slice(&value.to_network());
```

Comment:
This is again allocating vec internally in to_network, which is not needed here at all.

So maybe here the best strategy is to split to_network into two parts, one that allocates a vector of the correct size, and one that then fills the vector with the desired bytes, and just call the second part here

so to_network() would look somewhat like:
```rust
 pub fn to_network(&self) -> Vec<u8> {
     let mut vec = Vec::with_capacity(self.ser_byte_len());
     self.to_network_inner(&mut vec);
     vec
}
```
then you can use to_network inner here

Reply:
Added all of the above the performance changed by 1-3%.

Reply:
looks good!